[Dist] Enable FSDP on CPU

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #112145
* #112144

Differential Revision: [D50688958](https://our.internmc.facebook.com/intern/diff/D50688958/)

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

rohan-varma committed:
[Dist] Enable FSDP on CPU



Differential Revision: [D50688958](https://our.internmc.facebook.com/intern/diff/D50688958/)

[ghstack-poisoned]

fegin reviewed (approved):
LGTM

wanchaol commented on torch/cpu/__init__.py:
```diff
@@ -55,8 +55,25 @@ class Stream:
+        pass
+
+
+class Event:
```
Let's add a comment to this class since it's technically a public API:

```
    """
    N.B. This class only exists to facilitate device-agnostic code
    """
```
Maybe also need to add a doc entry for this

rohan-varma committed:
Update on "[Dist] Enable FSDP on CPU"


Differential Revision: [D50688958](https://our.internmc.facebook.com/intern/diff/D50688958/)

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

rohan-varma committed:
Update on "[Dist] Enable FSDP on CPU"


Differential Revision: [D50688958](https://our.internmc.facebook.com/intern/diff/D50688958/)

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

rohan-varma commented:
@pytorchbot merge -f "Ci failures are flaky"

rohan-varma commented:
@awgu I mostly landed it for testing and because some users were asking to use it in their testing use cases, but what do you mean by numerically correct? While we haven't done any correctness testing, none of the logic in FSDP is changed?
